### ISSUE TITLE
IO redesign

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#222]
We have agreed to reorder inputs in python in order incorporate domains more easily/generically and to prepare for the public release.  The plan involves splitting the input into more logical blocks, and deferring anything but building inputs until all of the inputs are read in.  This activity is related to a number of other issues how advanced and deprecated modes are handled.  Specific issues have been documented in a number of other items, including #174 #201 #214 #215

**Comment by kslong** [comment]
This issue is being dealt with in the branch called inputs.

A first task is to remove any actual work other than the gathering of inputs from the io section of the code.  As far as I can determine from the main part of the program.  There is nothing done other than gather inputs, except:

reading the atomic data
reading in the hydro model, when this is an option
reading the windsave model from previous runs.  

If anything else happens, it must be happening in the various subroutines that are called for getting inputs.   Reading the windsave model is required in situations where we want to use values from and old run.  The atomic data is required for setting up pointers properly.  So only the hydro model read can be postponed.

**Comment by kslong** [comment]
Python has 3 basic operating modalities that need to be considered in setting up the inputs

A new run of a program
A restart of a program
A run that begins with an old wind model.

These different modes require different inputs from the .pf file (or should)

For the normal mode everything has to come from the .pf file or the command line.

For a restart of a program, nothing really should change, with the reasonable exception of the number of cycles in each mode the number of photons per cycle.

For a run that begins with an old wind model, the situation is more complicated.  In general anything that affects the structure of the wind itself has to be maintained, but things having to do with radiation sources might not.   One of the ideas behind using an old model would be to speed convergence, so clearly one would like to be able to start a new model with a different effective temperature for the star.  However, if you increase the size of the star, that would change the setup of the wind.  

I think this is too complicated to prevent folks from doing stupid things.

Here is a file showing what one currently is asked for when you start a run from an old wind file.

[previous.txt](https://github.com/agnwinds/python/files/489811/previous.txt)

**Comment by kslong** [comment]
An  issue with our io redesign is that currently we have defined variables called

geo.wind_domain_number
geo.atmos_domain_number

If our goal is to generalize domains these need to be removed.  One of the places these variables are used are in @Higginbottom  's hydro models.

-- I have removed these variables in the latest version on the inputs branch.  I've created a new variable geo.hydro_domain_number to handle the zeus runs.  At this point, I have not reordered any inputs or allowed for multiple domains (other than a disk atmosphere) -- ksl

**Comment by kslong** [comment]
On branch inputs, I have significantly changed the way rdpar works, and in particular it should now accommodate simple reordering of the inputs, without dropping one into interactive mode.  In addition, if there is a missing keyword in the parameter file, it should request that keyword from the command line and then continue using values in the .pf file.  A few other features are described in the git log for inputs.  I would appreciate @jhmatthews @Higginbottom and @smangham giving this a try, and letting me know if this is an improvement.  (Note - branch inputs also handles multiple domains generically so if you want to play around with say a fast and a slow wind region, this is the branch to use.)

**Comment by jhmatthews** [comment]
Knox (@kslong) I've had a look at this. It looks good! I tried messing things around and the code seemed to deal with it well. The synonyms file seems sensible as well and is worth maintaining. 

We still need to sort out some ordering things, and I think the disk atmosphere command should probably be a wind domain and atmosphere should thus become one of the variables in wind type, but this looks like a good start.

**Comment by kslong** [comment]
I agree with James’ last point that a disk atmosphere is just a wind domain.  This is now an easy change to make.

**Comment by kslong** [comment]
There are some small issues that remain here, some indicated in #228.  Another issue  is that previously when one was entering inputs and decided one wanted to start over, one could quit with Ctrl D. This is no longer the case.  The only way to stop is with Control C.  This may be OK but it is a change in the way the routine functions and it is not obvious that his change is desirable.

**Comment by kslong** [comment]
There still are issues with IO order, such as discussed in #214.  I am beginning to think about this again, in my spare time.

One of the reasons we have problems has to do with the fact that there are basically 3 use cases for Python usage

1.  Restarts -- where we are completing a calculation and we do not want anything to change.  We simply want to finish what we planned to do
2. Previous model -- In this case we are beginning with a previous model, but we want more control over variables that do not directly affect the geometry.  For example, we could simply want to change the T of the star, or the angles over which something is extracted.  
3. New model - The normal case where we are starting from scratch

Item 2 is what makes everything complicated because it is not always obvious what one can change.  Clearly one cannot change the wind gridding, but could one in fact change to ionization approach ... probably, or at least in some cases.  You probably cannot switch from simple ions to macro atoms, or in general change the atomic data files, since we dynamically allocate the data arrays.  

Aside: The way the inputs are read into python is into python in the same order as the above. That is we check if we want to continue an uncompleted model before checking if we want to extend a previous model, before we start a completely new one.

**Comment by kslong** [comment]
@jhmatthews  @Higginbottom 

Please comment on these, one for a CV and one for an AGN.  Everything is not complete yet but you should get the drift.  (In particular, I have not dealt completely with the BL, and I will move the wind_radiation down a bit, and add more comments to the file.)

[agn.out.pf.txt](https://github.com/agnwinds/python/files/1396573/agn.out.pf.txt)
[cv.out.pf.txt](https://github.com/agnwinds/python/files/1396574/cv.out.pf.txt)

**Comment by kslong** [comment]
@jhmatthews  @Higginbottom @smangham 

I have completed a set of changes designed to make the IO more readable, and merged this back into dev as version 82b.  Old pf files should still work, though at some point everyone should migrate to the new versions.   You can see what this has done to the input files for say agn.pf, by looking the  agn.out.pf.   Agn.out.pf, etc, is the output file one should look at going fowar.

While I don't believe I changed the functionality anywhere, I did move code around.  In particularly, the functionality that was in get_radiation_sources has been redistributed.  So you should check that the functionality of the code has not changed, the worry being that in changing the order of the input variables, something hidden is not initialized properly.   I've checked the files the regression directory, but this is one of those times it would be useful to have had more diverse regression examples (See #304)

**Comment by kslong** [comment]
Although this, at some level remains an ongoing issue, we have made a lot of progress, and understand the path we are on, so I am closing this.
